[FLINK-18722][mesos] Migrate MesosResourceManager to the new MesosResourceManagerDriver

xintongsong opened the pull request:
## What is the purpose of the change

This PR migrates the Mesos deployment from the legacy `MesosResourceManager` to the new unified `ActiveResourceManager`.

## Brief change log

- 45c99aa2095e0eb7c826e0977cc7700a71c0f3ab..d1b22baf2b257bb71fe23ea633b985a965c940c1: Code clean-ups, in preparation for testability, including separation of concerns and interface abstractions.
- a2906880e06f7d2621ba91651637e2eb00f32503..797ad2e479f7ba4877ce94fb759e60df08bd0fce: Update interfaces between `ActiveResourceManager` and `ResourceManagerDriver`, w.r.t. requirements from the Mesos deployment.
- 254b849418cf257c49c361c80fda41f4803d38b0..92a61e033b1931f8fd67d9a468d38dcc5c8f410a: Main commit introducing `MesosResourceManagerDriver`. Testing codes are in separated commits for review, and should be squashed into the main commit on merging.
- 69af2b4e5264416bcc1a9edb35dae89102cf82b3: Switch to the new `MesosResourceManagerDriver` and clean up legacy codes.

## Verifying this change

- Add `MesosResourceManagerDriverTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

xintongsong commented:
cc @tillrohrmann 
Could you help take a look at this? 
The testing codes are in separated commits for review, which should be squashed before merging.

XComp commented on flink-mesos/src/test/java/org/apache/flink/mesos/runtime/clusterframework/MesosResourceManagerDriverTest.java:
```diff
@@ -0,0 +1,361 @@
+	private static final String SLAVE_HOST = "slave-host";
+
+	@Test
+	public void testAcceptOffers() throws Exception {
```
Out of curiosity: What's the difference between `MesosResourceManagerDriverTest::testAcceptOffers` and `ResourceManagerDriverTestBase::testRequestResource`. Based on the code it looks like both are doing the same thing. 🤔 

Just trying to answer my questions myself: Is `MesosResourceManagerDriverTest::testAcceptOffers` testing whether `requestResource` is coming back successfully whereas `ResourceManagerDriverTestBase::testRequestResource` tests whether we reach the Mesos backend via `requestResource`?

xintongsong replied:
`MesosResourceManagerDriverTest::testAcceptOffers` is testing that when resource offer is received from Mesos, `MesosResourceManagerDriver` will properly notify `ActiveResourceManager` about that, by completing the request resource future.

A `ResourceManagerDriver` works between two components: `ActiveResourceManager` and the external resource manager, thus its behaviors should be verified bidirectionally. Practically, test cases in `ResourceManagerDriverTestBase` verify that events from `ActiveResourceManager` are properly converted and forwarded to the external resource manager, while test cases in the deployment specific test class verify that events from the external resource manager are properly converted and forwarded to the `ActiveResourceManager`.

XComp replied:
Thanks for clarification.

XComp commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosResourceManagerDriver.java:
```diff
@@ -739,172 +591,123 @@ static ResourceID extractResourceID(Protos.TaskID taskId) {
+	@VisibleForTesting
+	void assertStateCleared() {
+		assert(workersInNew.isEmpty());
+		assert(requestResourceFutures.isEmpty());
```
Would it make sense to also check that the actors are cleared (i.e. `MesosResourceManagerDriver::stopSupportingActorsAsync` was performed properly)?

xintongsong replied:
This is a production method. It is expected to be called after the actors are created and before any new worker is requested/added.

I guess you mean verifying actors being cleared in `MesosResourceManagerDriverTest::testClearStateAfterRevokeLeadership`? That makes sense to me. We can verify wether `stopActor` are called for all the actors.

XComp replied:
Exactly. I came across this code segment when going through `MesosResourceManagerDriverTest::testClearStateAfterRevokeLeadership`.

XComp reviewed (approved):
Hi @xintongsong , thanks again for working on the refactoring. I just found minor stuff. @tillrohrmann will check as well after he's back from vacation.

xintongsong commented:
Thanks for the review, @XComp. Nice catches.
I've addressed your comments.

xintongsong commented:
Rebase to resolve conflicts.

tillrohrmann reviewed (commented):
Thanks for creating this PR @xintongsong and the review @XComp. The changes look already very good. I have two comments:

1. I think we should not get rid of the `ioExecutor` because then we can no longer control in which thread pool the worker recovery happens.
2. I think we don't need to run unfenced runnables. This is unrelated to this PR but if we change this, then we don't have to change the executor of the `ResourceManagerDriver`.

xintongsong commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosResourceManagerDriver.java:
```diff
@@ -0,0 +1,713 @@
+	private CompletableFuture<List<MesosWorkerStore.Worker>> getWorkersAsync() {
+		// if this resource manager is recovering from failure,
+		// then some worker tasks are most likely still alive and we can re-obtain them
+		return CompletableFuture.supplyAsync(() -> {
+			try {
+				final List<MesosWorkerStore.Worker> tasksFromPreviousAttempts = workerStore.recoverWorkers();
+				for (final MesosWorkerStore.Worker worker : tasksFromPreviousAttempts) {
+					if (worker.state() == MesosWorkerStore.WorkerState.New) {
+						// remove new workers because allocation requests are transient
+						workerStore.removeWorker(worker.taskID());
+					}
+				}
+				return tasksFromPreviousAttempts;
+			} catch (final Exception e) {
+				throw new CompletionException(new ResourceManagerException(e));
+			}
+		});
```
True. Robert has introduced `ioExecutor` to resource managers in FLINK-19241. I think we can pass it from `ActiveResourceManager` in `ResourceManagerDriver#initialize`.

xintongsong commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosResourceManagerDriver.java:
```diff
@@ -0,0 +1,713 @@
+
+		@Override
+		public void disconnected(SchedulerDriver driver) {
+			getMainThreadExecutor().executeWithoutFencing(() -> MesosResourceManagerDriver.this.disconnected(new Disconnected()));
```
Looks like changing `executeWithoutFencing` to `execute` does not break any test cases.

I can try to provide an implementation that handles disconnected messages with fencing. Could you tell me how to run it on `flink-jepsen`?

tillrohrmann replied:
I think you cannot do it alone because we have the Jepsen source code for running it on Travis in a private repo.

I started the jepsen run with this branch: https://github.com/tillrohrmann/flink/tree/disableUnfencedExecutionMesos

We can see the CI run here: https://travis-ci.com/github/dataArtisans/flink-jepsen-ci/builds/189052564

Once the run passes I will open a PR for the changes and then merge it. You can already rebase this PR on my branch if you want.

xintongsong commented:
@tillrohrmann, thanks for your review. I've addressed the comments.
Could you help me with some instructions on how to test the changes on `flink-jepsen`?

tillrohrmann commented:
I've started the Jepsen run based on https://github.com/tillrohrmann/flink/tree/disableUnfencedExecutionMesos. The Jepsen run can be seen here: https://travis-ci.com/github/dataArtisans/flink-jepsen-ci/builds/189052564. I will open a PR for the removal of the run without fencing and merge it once Jepsen gives green light.

tillrohrmann commented:
For future reference: The PR for removing the `runAsyncWithoutFencing` can be found here: #13573.

xintongsong commented:
@tillrohrmann, thanks for the Jepsen run and the #13573.
I do not have the access for opening the travis link. Would you please comment the Jepsen run result.
Meantime, I'll start rebasing this PR.

xintongsong commented:
Done rebasing.

tillrohrmann reviewed (approved):
Thanks for resolving my comments @xintongsong. The changes look good to me. +1 for merging once AZP gives green light.

xintongsong commented:
Thanks for the review, @tillrohrmann. Merging this.
